fixes Jandex index is not complete, missing: java.lang.Comparable #1580

machi1990 opened the pull request:
See #1580

machi1990 committed:
fixes Jandex index is not complete, missing: java.lang.Comparable #1580

gsmet commented on extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/JpaJandexScavenger.java:
```diff
@@ -121,70 +122,86 @@ private static void enlistExplicitClasses(IndexView index, JpaEntitiesBuildItem
+            domainObjectCollector.addEntity(targetDotName.toString());
         }
+
+        throwIfUnIndexedClassesNotEmpty(unIndexedClasses);
```
I usually avoid doing that as you completely lose the flow.,

I would do:
```
if (!unIndexedClasses.isEmpty()) {
    throw unindexedClassesException(unindexedClasses.isEmpty);
}
```

machi1990 replied:
Done. Moved into `discoverModelAndRegisterForReflection` with the proposed change.

gsmet commented on extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/JpaJandexScavenger.java:
```diff
@@ -121,70 +122,86 @@ private static void enlistExplicitClasses(IndexView index, JpaEntitiesBuildItem
+            return;
+        }
+
+        Set<DotName> unIndexedClasses = new HashSet<>();
```
I would make it `unindexedClasses` as in `ReflectiveHierarchyStep`. Let's be consistent considering we follow the same pattern.

machi1990 replied:
Sold. :-)

gsmet commented on extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/JpaJandexScavenger.java:
```diff
@@ -121,70 +122,86 @@ private static void enlistExplicitClasses(IndexView index, JpaEntitiesBuildItem
+                    unIndexedClasses);
         }
+
+        throwIfUnIndexedClassesNotEmpty(unIndexedClasses);
```
I don't think the pattern is right. You should collect all of them and throw only one exception. Here you collect 2 different lists. For me, this should be done only once in `discoverModelAndRegisterForReflection`.

machi1990 replied:
Good catch.

gsmet commented on extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/JpaJandexScavenger.java:
```diff
@@ -196,13 +213,29 @@ private static void addClassHierarchyToReflectiveList(IndexView index, JpaEntiti
         }
 
-        //Capture this one (for various needs: Reflective access enablement, Hibernate enhancement, JPA Template)
+        //Capture this one (for various needs: Reflective access enabling, Hibernate enhancement, JPA Template)
```
That was correct. It should be reverted.

machi1990 replied:
Done.

gsmet reviewed (changes requested):
@machi1990 thanks. Looks in line with what we discussed today.

I added some comments inline.

stuartwdouglas commented on extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/JpaJandexScavenger.java:
```diff
@@ -102,16 +104,26 @@ public JpaEntitiesBuildItem discoverModelAndRegisterForReflection() throws IOExc
+        if (!unindexedClasses.isEmpty()) {
+            final String unindexedClassesErrorMessage = unindexedClasses.stream().map(d -> "\t- " + d + "\n")
+                    .collect(Collectors.joining());
+            throw new ConfigurationError(
```
I don't know if this should cause the augmentation to fail. This only really affects native mode, so you might prevent an otherwise valid app from running in JVM mode, when the end user has no intention of actually using native mode.

machi1990 replied:
@stuartwdouglas good remark. We should probably log a warning message instead of throwing?

gsmet replied:
I know @dmlloyd was thinking about making JVM and native as close as possible so that when your app is working in JVM mode, you can be fairly confident it works in native mode.

BTW, it does not only register the classes for reflection, it also collects the list of domain objects.

machi1990 replied:
I saw the [discussion](https://github.com/quarkusio/quarkus/issues/1605#issuecomment-475217200) @gsmet was referring to. As a dev, I'd like the kind of consistency and assurance that "it works everywhere" brings in my development process.

gsmet commented on extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/JpaJandexScavenger.java:
```diff
@@ -102,16 +104,26 @@ public JpaEntitiesBuildItem discoverModelAndRegisterForReflection() throws IOExc
-                addClassHierarchyToReflectiveList(index, domainObjectCollector, enumTypeCollector, dotName);
+                addClassHierarchyToReflectiveList(index, domainObjectCollector, enumTypeCollector, dotName, unindexedClasses);
             } else {
                 // We do lipstick service by manually adding explicitly the <class> reference but not navigating the hierarchy
```
Shouldn't we use the same pattern here and add the class to the `unindexedClasses`?

machi1990 replied:
@gsmet I was tempted to do so but the comment made it look like something else here is done: add the class anyway and do not scan the whole hierarchy.

machi1990 replied:
It is the same [here](https://github.com/quarkusio/quarkus/blob/6b5bc32175d08538523dca2196bc91bd8a726dae/extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/JpaJandexScavenger.java#L133). 

I was also thinking if there is a way to document on how to  "index" custom dependencies as suggested by this [line](https://github.com/quarkusio/quarkus/blob/6b5bc32175d08538523dca2196bc91bd8a726dae/extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/JpaJandexScavenger.java#L131) but with working link.

Sanne commented:
@gsmet is this one good enough now?

gsmet commented:
@Sanne I still have to work on it. Planned for tonight or tomorrow morning. Before the release, anyway.

gsmet committed:
Create a Jandex index for the Panache classes

We need the entity hierarchy to be indexed so PanacheEntity needs to be
indexed.

gsmet committed:
Polish the warn message for ReflectiveHierarchyStep

gsmet committed:
Throw an error in any case when trying to enlist the JPA classes

It is not only necessary for reflection, it's also necessary for
proper enhancement so we really need all the hierarchy to be property
indexed.

gsmet commented:
I added a few commits. I was a bit more aggressive as I really think we should throw an error if everything is not indexed. It definitely can lead to weird issues, classes not being properly enhanced and so on.

machi1990 commented:
@gsmet seen them. Looks good to me. Thanks.

Sanne commented:
thanks all! merged
